### PR TITLE
[libffi] Update to v3.3 release

### PR DESCRIPTION
**Pull request by driver1998**
__Update:__ v3.3 finally released(3e249f826e77c57f4a85be746dcfd1f891ab90f4), hopefully it will just work.

~~This is a port of libffi master as of Apr 10 (d1e9b4b), with both https://github.com/libffi/libffi/pull/483 (arm64 support) and https://github.com/libffi/libffi/pull/477 (arm32 support).~~

~~This port is work in progress.~~

### CODE REVIEW

**Comment by Rastaban** [comment]
Thanks for all of the new PRs!  With this one I see a new build failure with libffi on osx and Linux.  Are you able to investigate those?  Also do you know when libffi is planning on releasing 3.3?  We prefer to stick with official releases rather than switching to an unofficial repro and using a hash.

**Comment by driver1998** [comment]
@Rastaban I agree with you about the official releases, this is currently just a temporary hack for testing & using glib on arm-windows / arm64-windows. That is also why it is marked as WIP.
As for the 3.3 release, I am not 100% sure about that though.

**Comment by driver1998** [comment]
I'll see what I can do to the Linux side of things soon. 

As for OS X, I don't have a Mac so sorry about that. (Maybe they are related)

**Comment by grdowns** [comment]
Hey @driver1998, here is the failure log for `x64-linux` if you're interested in seeing it. Note that the debug installation links fine while the release installation fails -- I might be able to take a deeper look at it at some point if you don't get the chance to. Thanks for all these PR's!
[install-x64-linux-rel-out.log](https://github.com/Microsoft/vcpkg/files/3109794/install-x64-linux-rel-out.log)

**Comment by driver1998** [comment]
批注 2019-04-28 013830
So it works on my end, Xubuntu 18.04 amd64.
Can you provide me a little bit more detail?

**Comment by driver1998** [comment]
With this commit I also updated to a more recent commit, and we are pretty close to main branch right now (waiting for one more PR to upstream).

**Comment by grdowns** [comment]
`x64-linux` seems to be building fine according to the CI now -- that might have been an error on my part. Thanks for confirming the results and giving us a status update, @driver1998. :smile:

**Comment by grdowns** [comment]
Hey @driver1998, could you please link the upstream pull requests so we can follow along with the progress?

**Comment by driver1998** [comment]
https://github.com/libffi/libffi/pull/490

**Comment by ras0219-msft** [comment]
Let's move forward with this by pulling the PR branch, since upstream hasn't looked at the PR in a month.

**Comment by driver1998** [comment]
Now https://github.com/libffi/libffi/pull/486 is merged, I am going to move to upstream master...

**Comment by driver1998** [comment]
https://github.com/libffi/libffi/pull/490 contains one more patch over https://github.com/libffi/libffi/pull/486, which is added as a patch to the port.

The new PR for this patch is https://github.com/libffi/libffi/pull/496.

**Comment by driver1998** [comment]
Now everything we need is in upstream master, only to wait for.an official 3.3 release...

**Comment by c72578** [comment]
> Now everything we need is in upstream master, only to wait for.an official 3.3 release...

In the meantime,  libffi 3.3 Release Candidate 1 has been released (2019-10-24). However, not the official 3.3 yet, but definitely worth testing.
https://github.com/libffi/libffi/releases/tag/v3.3-rc1

**Comment by qz3fwd** [comment]
Adding the more recent version which supports arm64 would be great. Last night I hacked together a portfile to build the arm64 lib using the msbuild function. I would prefer the port be updated rather than my hack.

**Comment by driver1998** [comment]
Rebased this branch to current vcpkg master, and upgrade to v3.3-rc2 (just released).

**Comment by driver1998** [comment]
Linux CI failed for some other reason, the build actually finished.

**Comment by driver1998** [comment]
@janisozaur:
> You probably didn't rebase properly, I saw some merge commits, perhaps you did a merge? Keeping such work in single PR allows everyone to follow discussion and personally I would urge you restore the previous PR with contents of this branch and continue working there. The least you could do would be to leave a comment in #6119, explaining why was it closed and where to look for continuation.

Sorry for the confusion, reopen this thread now.

**Comment by janisozaur** [comment]
@driver1998 could you please fixup contents of this PR? By that I mean cleaning the merges, so it doesn't say you're trying to merge 746 commits totalling +36,652 −19,500. Ideally that would mean rebasing, but something as trivial as applying https://patch-diff.githubusercontent.com/raw/microsoft/vcpkg/pull/8878.patch on top of current master and pushing that as driver1998:libffi should suffice as well.

Note: this is just my personal opinion, I'm only contributor here, but I would surely not allow any such PR into projects I maintain.

**Comment by driver1998** [comment]
@janisozaur looks like the last merge messed up all the things, now it is much better.

**Comment by janisozaur** [comment]
Many thanks!

**Comment by driver1998** [comment]
Update: v3.3 finally released, hopefully it will just work.

**Comment by driver1998** [comment]
I already updated this port to build the official 3.3 release, and it works fine. Please merge this PR.

**Comment by grdowns** [comment]
Hey @driver1998! The PR LGTM. We're currently having some problems with our CI baseline and the fix will be in very soon. I've removed the tags and I will prioritize merging this PR as soon as the changes go live. Thanks for your patience and thanks for working on this longstanding changeset!

File: ports/libffi/CMakeLists.txt [hunk 1]
```diff
@@ -35,43 +39,86 @@ else()
...
 
 include_directories(${CMAKE_BINARY_DIR}/include)
 include_directories(${CMAKE_BINARY_DIR})
 include_directories(include)
 
-add_definitions(-DHAVE_CONFIG_H)
 add_definitions(-DFFI_BUILDING)
+if(BUILD_SHARED_LIBS AND WIN32)
+    add_definitions(-DFFI_BUILDING_DLL)
+    set(CMAKE_WINDOWS_EXPORT_ALL_SYMBOLS ON)
```
Comment (dan-shaw) [inline ports/libffi/CMakeLists.txt]: We don't want to use `CMAKE_WINDOWS_EXPORT_ALL_SYMBOLS`, see https://github.com/microsoft/vcpkg/blob/master/docs/maintainers/maintainer-guide.md#do-not-add-cmake_windows_export_all_symbols

Comment (driver1998) [reply ports/libffi/CMakeLists.txt]: Seems to work fine without it even with shared build, let's see how it goes.

File: ports/libffi/portfile.cmake [hunk 2]
```diff
@@ -1,19 +1,11 @@
 include(vcpkg_common_functions)
```
Comment (NancyLi1013) [inline ports/libffi/portfile.cmake]: `include(vcpkg_common_functions)` is no longer needed.
Could you please remove this line?

File: ports/libffi/portfile.cmake [hunk 3]
```diff
@@ -32,14 +24,15 @@ vcpkg_install_cmake()
...
+"   *know* they are going to link with the static library.  */
 
 #define FFI_BUILDING
-
 " FFI_H "${FFI_H}")
 file(WRITE ${CURRENT_PACKAGES_DIR}/include/ffi.h "${FFI_H}")
+endif()
 
 file(COPY ${SOURCE_PATH}/LICENSE DESTINATION ${CURRENT_PACKAGES_DIR}/share/libffi)
 file(RENAME ${CURRENT_PACKAGES_DIR}/share/libffi/LICENSE ${CURRENT_PACKAGES_DIR}/share/libffi/copyright)
```
Comment (NancyLi1013) [inline ports/libffi/portfile.cmake]: Could you update 
```
file(COPY ${SOURCE_PATH}/LICENSE DESTINATION ${CURRENT_PACKAGES_DIR}/share/libffi)
file(RENAME ${CURRENT_PACKAGES_DIR}/share/libffi/LICENSE ${CURRENT_PACKAGES_DIR}/share/libffi/copyright)
```
as
`file(INSTALL ${SOURCE_PATH}/LICENSE DESTINATION ${CURRENT_PACKAGES_DIR}/share/${PORT} RENAME copyright)`?

**Review by NancyLi1013 — CHANGES_REQUESTED** [review]

**Comment by PhoebeHui** [comment]
@driver1998, could you please resolve the conflicts?

**Comment by driver1998** [comment]
Resolved.

**Comment by PhoebeHui** [comment]
@driver1998, thanks for updates!

The CI failed with following failures, could you please take a look? if this is expected, could you please update them to ci.baseline.txt file?

##[error] REGRESSION: libnice:arm64-windows
##[error] REGRESSION: gdk-pixbuf:arm64-windows
##[error] REGRESSION: pangomm:arm64-windows
##[error] PASSING, REMOVE FROM FAIL LIST: libffi:arm64-windows (C:\vsts\_work\3\s\scripts\ci.baseline.txt)
##[error] REGRESSION: atk:arm64-windows

**Comment by driver1998** [comment]
Should be resolved now.

File: ports/pangomm/portfile.cmake [hunk 4]
```diff
@@ -1,5 +1,9 @@
 include(vcpkg_common_functions)
```
Comment (NancyLi1013) [inline ports/pangomm/portfile.cmake]: `include(vcpkg_common_functions)` is no longer needed.
Could you please remove this?

File: ports/pangomm/portfile.cmake [hunk 5]
```diff
@@ -1,5 +1,9 @@
...
 
+if(VCPKG_TARGET_ARCHITECTURE STREQUAL "arm" OR VCPKG_TARGET_ARCHITECTURE STREQUAL "arm64")
+    message(FATAL_ERROR "This port does not currently support architecture: ${VCPKG_TARGET_ARCHITECTURE}")
+endif()
+
```
Comment (NancyLi1013) [inline ports/pangomm/portfile.cmake]: These lines can be changed as `vcpkg_fail_port_install(ON_ARCH "arm" "arm64")`.
Could you please update this?

File: ports/pangomm/portfile.cmake [hunk 6]
```diff
@@ -1,5 +1,9 @@
...
 vcpkg_download_distfile(ARCHIVE
     URLS "http://ftp.gnome.org/pub/GNOME/sources/pangomm/2.40/pangomm-2.40.1.tar.xz"
     FILENAME "pangomm-2.40.1.tar.xz"
```
Comment (NancyLi1013) [inline ports/pangomm/portfile.cmake]: Could you please update 

`file(INSTALL ${SOURCE_PATH}/COPYING DESTINATION ${CURRENT_PACKAGES_DIR}/share/pangomm RENAME copyright)`
as

`file(INSTALL ${SOURCE_PATH}/COPYING DESTINATION ${CURRENT_PACKAGES_DIR}/share/${PORT} RENAME copyright)`?

**Review by NancyLi1013 — CHANGES_REQUESTED** [review]

**Review by NancyLi1013 — APPROVED** [review]

**Review by PhoebeHui — APPROVED** [review]
LGTM, thanks for the PR!

**Review by ras0219-msft — APPROVED** [review]

**Comment by ras0219-msft** [comment]
Thanks for the PR!

**Pull request merged**